pkg/kubelet/container_manager_linux.go

```diff
@@ -34,73 +36,132 @@ import (
+		cont := newSystemContainer(dockerDaemonContainerName)
+		cont.ensureStateFunc = func(manager *fs.Manager) error {
+			return ensureDockerInContainer(-900, createManager(dockerDaemonContainerName))
+		}
+		systemContainers = append(systemContainers, cont)
 	}
 
-	return &containerManagerImpl{
-		useDockerContainer: dockerDaemonContainer != "",
-		useSystemContainer: systemContainer != "",
-		dockerOomScoreAdj:  -900,
-		dockerContainer: fs.Manager{
-			Cgroups: &configs.Cgroup{
-				Name:            dockerDaemonContainer,
-				AllowAllDevices: true,
-			},
-		},
-		systemContainer: fs.Manager{
-			Cgroups: &configs.Cgroup{
-				Name:            systemContainer,
-				AllowAllDevices: true,
-			},
-		},
-		rootContainer: fs.Manager{
+	if systemContainerName != "" {
+		if systemContainerName == "/" {
+			return nil, fmt.Errorf("system container cannot be root (\"/\")")
+		}
+
+		rootContainer := &fs.Manager{
 			Cgroups: &configs.Cgroup{
 				Name: "/",
 			},
-		},
+		}
+		manager := createManager(systemContainerName)
+
+		err := ensureSystemContainer(rootContainer, manager)
+		if err != nil {
+			return nil, err
+		}
+		systemContainers = append(systemContainers, newSystemContainer(systemContainerName))
+	}
+
+	if kubeletContainerName != "" {
+		systemContainers = append(systemContainers, newSystemContainer(kubeletContainerName))
+	}
+
+	// TODO(vmarmol): Add Kube-proxy container.
+
+	return &containerManagerImpl{
+		systemContainers: systemContainers,
 	}, nil
 }
 
+// Create a cgroup container manager.
+func createManager(containerName string) *fs.Manager {
+	return &fs.Manager{
+		Cgroups: &configs.Cgroup{
+			Name:            containerName,
+			AllowAllDevices: true,
+		},
+	}
+}
+
 func (cm *containerManagerImpl) Start() error {
-	if cm.useSystemContainer {
-		err := cm.ensureSystemContainer()
-		if err != nil {
-			return err
+	// Don't run a background thread if there are no ensureStateFuncs.
+	numEnsureStateFuncs := 0
+	for _, cont := range cm.systemContainers {
+		if cont.ensureStateFunc != nil {
+			numEnsureStateFuncs++
 		}
 	}
-	if cm.useDockerContainer {
-		go util.Until(func() {
-			err := cm.ensureDockerInContainer()
-			if err != nil {
-				glog.Warningf("[ContainerManager] Failed to ensure Docker is in a container: %v", err)
-			}
-		}, time.Minute, util.NeverStop)
+	if numEnsureStateFuncs == 0 {
+		return nil
 	}
+
+	// Run ensure state functions every minute.
+	go util.Until(func() {
+		for _, cont := range cm.systemContainers {
+			if cont.ensureStateFunc != nil {
+				err := cont.ensureStateFunc(cont.manager)
+				glog.Warningf("[ContainerManager] Failed to ensure state of %q: %v", cont.name, err)
+			}
+		}
+	}, time.Minute, util.NeverStop)
+
 	return nil
 }
 
+func (cm *containerManagerImpl) SystemContainersReservation() api.ResourceList {
+	cpuReservation := int64(0)
+
+	// Sum up resources of all external containers.
+	for _, cont := range cm.systemContainers {
+		cpuReservation += cont.cpuMillicores
```

Comment:
right now, we don't set cpuMillicores. Is it going to return 0 or 1000 for every container?